spring-data-jpa/src/main/java/org/springframework/data/jpa/repository/query/JpqlQueryTransformer.java

```diff
@@ -0,0 +1,2437 @@
+/*
+ * Copyright 2022-2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.jpa.repository.query;
+
+import static org.springframework.data.jpa.repository.query.QueryParsingToken.*;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.springframework.data.domain.Sort;
+import org.springframework.lang.Nullable;
+
+/**
+ * An ANTLR {@link org.antlr.v4.runtime.tree.ParseTreeVisitor} that transforms a parsed JPQL query.
+ *
+ * @author Greg Turnquist
+ * @since 3.1
+ */
+class JpqlQueryTransformer extends JpqlBaseVisitor<List<QueryParsingToken>> {
```

Comment:
This and the HQL transformer require additional work.

1. The many occurrences of `getText()` (e.g. `new QueryParsingToken(ctx.FROM().getText(), true)`) indicate that the code isn't ideal. We're creating a lot of objects for mostly static components that we could cache. Also, regular `getText()` calls could be abstracted away to avoid duplications.
2. This class combines query parsing/rendering along with our domain-specific tweaks to introduce sorting, count query projections. Ideally, we can encapsulate all base-work that is necessary for query rendering into a base transformer class and have transformers for specific tasks (count query derivation, sorting, …). To put things into perspective: `QueryParser` has individual methods for each use case, while this class combines all use-cases.
3. HQL and JPQL share a lot of common ground. It isn't ideal that we have so many duplications. Also, when comparing both files side-by-side, it's impossible to put both files next to each other to determine differences. Going forward, this will be the method to detect, if we apply a change to one, that we need to apply the same change to the other transformer as well. We need to fix that. First step would be to align method orders (e.g. that the HQL `visitJoinType` method is at the same place where the JPQL `visitJoin_spec` method is). While this is purely a style aspect from the technical perspective, it is something that decides over how difficult it becomes to maintain parsers.